rosutil/symdebug.rkt

```diff
@@ -0,0 +1,58 @@
+#lang racket
+
+(require (prefix-in r: rosette/safe))
+
+; Utilities for inspecting symbolic expressions in Yosys module state
+
+(provide term-length analyze-terms wide-write)
+
+; helper function that returns flattened length of an arbitrarily nested list
+(define (flattened-length l)
+  (if (not (list? l))
+      1
+      (if (empty? l)
+          0
+          (+ (flattened-length (first l))
+             (flattened-length (rest l))))))
+
+; returns "length" of (the possibly symbolic) t, depending on type of t:
+; vector? returns sum of term-length of each element
+; [symbolic] expression? returns flattened-length of t
+; [symbolic] constant? returns 1
+; other (i.e., concrete value) returns 0
+(define (term-length t)
+  (if (vector? t)
+      (foldl + 0 (map term-length (vector->list t)))
+      (if (r:expression? t)
+          (flattened-length (r:term->datum t))
+          (if (r:term? t)
+              1
+              0))))
+
+; analyze each element in state that is accessed via state-getters
+; state-getters is same construction as the argument of the same name to
+; verify-deterministic-start in rtl/shiva
+;
+; returns number of terms with length >= threshold
+; additionally outputs information about long state components if output? is #t
+(define (analyze-terms state state-getters
```

Comment:
This seems useful!

I wonder how the static threshold works if we use the `value-size` thing.

What exactly were you using this for? If you were using it to find suspiciously large terms, I wonder if some functionality that identifies which terms are growing would be useful.

Reply:
I was indeed using it to find suspiciously large terms. I thought about adding some functionality like that, but didn't end up needing it for what I was working on at the moment. I'd be interested in building something like that off of this PR at some point though.